### PR TITLE
kubernetes: Delete session tokens on logout

### PR DESCRIPTION
**Pull request by petervo**
We will need to back port the shell changes once we agree on the best way to do it.

### CODE REVIEW

**Comment by stefwalter** [comment]
What bug does this actually fix? Neither the pull request or the commits have that info. And what happens when the browser closes? What should happen?

**Comment by petervo** [comment]
Working on the logout test failures

**Comment by petervo** [comment]
> What bug does this actually fix? Neither the pull request or the commits have that info. And what happens when the browser closes? What should happen?

Openshift wants us to delete the token on logout. https://bugzilla.redhat.com/show_bug.cgi?id=1441950. I'd rather do this on the front end, but it's tricky because everything is in iframes.

When the browser closes nothing happens. I think that is the correct behavior and matches what the openshift web console does.

File: pkg/kubernetes/scripts/kube-client-cockpit.js [hunk 1]
```diff
@@ -802,6 +802,48 @@
...
+                token = parts[parts.length-1];
+            }
+
+            function setupClearTokenOnHint() {
+                cockpit.transport.filter(function(message, channel, control) {
+                    var req;
+                    if (control && control.command == "hint" && control.logout) {
+                        /* config retrieved from kubectl? */
+                        if (info.type == "sessionData" && token) {
+                            req = new CockpitKubeRequest("DELETE", path+token, "");
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/kube-client-cockpit.js]: How is this not racy? The window is about to be unloaded.

**Comment by stefwalter** [comment]
It seems like this behavior is only necessary when the registry console or kube dashboard is the only thing that was displayed. In other words, this is a kube dashboard.

In addition, did you consider just running the DELETE command in the ```cockpit-kube-auth``` oauth authorize command when the session is done? It seems to be specific to this Cockpit authorize command. The ```cockpit-session``` command also does cleanup after a local session.

**Comment by petervo** [comment]
> It seems like this behavior is only necessary when the registry console or kube dashboard is the only thing that was displayed. In other words, this is a kube dashboard.

>In addition, did you consider just running the DELETE command in the cockpit-kube-auth oauth authorize command when the session is done? It seems to be specific to this Cockpit authorize command. The cockpit-session command also does cleanup after a local session.

Yes, the reason I didn't go with a backend solution was that I though that we may need something similar for other dashboards in the future. Having frontend per-iframe logout code seems like the sort of that may be useful. But we can add a fork-exec to the cockpit-kube-auth command if we are sure that's the only place we are going to want it. Fork-exec is a little weird in go but we should be able to make it work.

**Comment by stefwalter** [comment]
> Yes, the reason I didn't go with a backend solution was that I though that we may need something similar for other dashboards in the future. Having frontend per-iframe logout code seems like the sort of that may be useful.

Are you sure? Most of the time we're a real Linux login session. Our authentication tokens should be stored in the user's session, and should share logic with the command line. This seems like the exception rather than the rule ... and the place for such exceptions is the authorize commands.

> Fork-exec is a little weird in go but we should be able to make it work.

Perhaps there's a good idiom for running a program and waiting until it's done? It seems like this should be routine in golang, no?

**Pull request merged**